### PR TITLE
ROX-13805: Add check-style to GHA

### PR DESCRIPTION
**Pull request by janisz**
## Description

Add `make style` to github actions workflow

## Testing Performed

CI

### CODE REVIEW

**Comment by gavin-stackrox** [comment]
I think this should be separate from builds i.e. a new workflow style.yaml. Though triggered by the same label. What do you think?

**Comment by janisz** [comment]
Make sense. How about caches should this job has it's own cache? It updates go mod cache and golangci-lint cache

**Comment by gavin-stackrox** [comment]
> Make sense. How about caches should this job has it's own cache? It updates go mod cache and golangci-lint cache

I think it should be fine to share the cache from the build steps but I'm not 100% clear what is cached :)

**Comment by janisz** [comment]
> I think it should be fine to share the cache from the build steps but I'm not 100% clear what is cached :)

It's OK but it's not optimal as cache will be populated by the first job and `make style` install additional deps in go modules and populate golang ci cache. I think this step should have it's own cache.

File: .github/actions/cache-ui-dependencies/action.yaml [hunk 1]
```diff
@@ -13,12 +13,15 @@ runs:
...
           ui/deps
+          /github/home/.cache/yarn/v6
           /github/home/.cache/Cypress
           /usr/local/share/.cache
           ui/node_modules
           ui/apps/platform/node_modules
           ui/packages/ui-components/node_modules
           ui/packages/tailwind-config/node_modules
-        key: npm-v2-${{ hashFiles(inputs.lockFile) }}
+        key: npm-v2--${{ hashFiles(inputs.lockFile) }}-${{ github.job }}
```
Comment (janisz) [inline .github/actions/cache-ui-dependencies/action.yaml:23]: @gavin-stackrox Why do we need monorepo.lock? Why not using glob 
```
${{ hashFiles('**/package.json') }}
```
or even
```
${{ hashFiles('**/yarn.lock') }}
```

**Comment by janisz** [comment]
@gavin-stackrox PTAL

**Comment by gavin-stackrox** [comment]
@janisz style-checks in OSCI are now under 20 mins: https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-stackrox-stackrox-master-style-checks?buildId=1575481918969876480

**Comment by gavin-stackrox** [comment]
[hit send too early]
@janisz style-checks in OSCI are now under 20 mins: https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-stackrox-stackrox-master-style-checks?buildId=1575481918969876480
So is there any real benefit to adding this?

**Comment by janisz** [comment]
@gavin-stackrox That's great news! Let's close this PR. Month ago it was much longer sometimes 50m https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-stackrox-stackrox-master-style-checks?buildId=1568344902184472576

How was it possible to get so much speed up on OSCI? Are other builds faster as well?

**Comment by gavin-stackrox** [comment]
@janisz the essential reason for the improvement is that the test image is mirrored into the CI cluster and the tests were changed to simpler 'single step' jobs so the ci-operator does not do as much image copying that was the source of a lot of the initial slowness. I did this for all the required jobs, so go-unit is also a lot quicker now. https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-stackrox-stackrox-master-go-unit-tests

Building though still is slow, so I think `turbo-build` shall live on unless that can get resolved. My best guess for building slowness is that it requires a few pods and allocating them is where OSCI can be arbitrary: https://prow.ci.openshift.org/job-history/gs/origin-ci-test/pr-logs/directory/pull-ci-stackrox-stackrox-master-images

**Comment by msugakov** [comment]
Tomek, do you mind to try rebasing to the recent master?

**Comment by janisz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2916** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/2916" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/2916?utm_source=stack-comment).

**Comment by janisz** [comment]
@msugakov @gavin-stackrox @porridge

**Review by gavin-stackrox — APPROVED** [review]
The code change looks good to me. Any performance improvement?

File: .github/workflows/style.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+jobs:
+
+  style-check:
+    if: |
+      (github.event.action == 'labeled' && github.event.label.name == 'turbo-build') ||
+      (github.event.action != 'labeled' && contains(github.event.pull_request.labels.*.name, 'turbo-build'))
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/stackrox-io/apollo-ci:stackrox-test-0.3.50
```
Comment (porridge) [inline .github/workflows/style.yaml:20]: How do we make sure going forward that we bump this here consistently with other places? :thinking:

File: .github/workflows/style.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
+name: Style
+on:
+  pull_request:
+    types:
+    - opened
+    - synchronize
+    - labeled
```
Comment (msugakov) [inline .github/workflows/style.yaml:7]: `reopened` is missing.

> By default, a workflow only runs when a pull_request event's activity type is opened, synchronize, or reopened.

From https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

File: .github/workflows/style.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+env:
+  ROX_PRODUCT_BRANDING: RHACS_BRANDING
+
+jobs:
+
+  style-check:
+    if: |
+      (github.event.action == 'labeled' && github.event.label.name == 'turbo-build') ||
+      (github.event.action != 'labeled' && contains(github.event.pull_request.labels.*.name, 'turbo-build'))
```
Comment (msugakov) [inline .github/workflows/style.yaml:17]: Why not drop this `if` given that GitHub Actions migration is allowed and confirmed by now?

File: .github/workflows/style.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      - uses: ./.github/actions/create-concatenated-ui-monorepo-lock
+
+      - name: Cache UI dependencies
+        uses: ./.github/actions/cache-ui-dependencies
+
+      - name: Cache QA Test dependencies
+        uses: ./.github/actions/cache-gradle-dependencies
+
+      - name: PR labels
+        uses: joerick/pr-labels-action@v1.0.6
```
Comment (msugakov) [inline .github/workflows/style.yaml:40]: I would suggest not add 3-rd party dependency for the task to determine if a label is present on PR.
Some googling suggests that `${{contains( github.event.pull_request.labels.*.name, 'My Label')}}` will provide an answer if `My Label` is present.

File: .github/workflows/style.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+          if [[ -n "$GITHUB_PR_LABEL_CI_RELEASE_BUILD" ]]; then
+            cci-export GOTAGS release
+          fi
+          echo "GOTAGS: ${GOTAGS:-}"
+
+      - name: Fetch UI deps
+        run: make -C ui deps
+
+      - name: make style
+        run: make style
```
Comment (msugakov) [inline .github/workflows/style.yaml:53]: Why not
```suggestion
        run: scripts/ci/jobs/style-checks.sh
```
?

File: .github/actions/cache-gradle-dependencies/action.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  using: composite
+  steps:
+    - name: Cache Gradle Dependencies
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.gradle/caches
+          ~/.gradle/wrapper
+          ~/.m2/repository
+        key: npm-v2-${{ hashFiles('**/*.gradle*', '**/gradle-wrapper.properties') }}-${{ github.job }}
```
Comment (msugakov) [inline .github/actions/cache-gradle-dependencies/action.yaml:13]: Why `npm-` or is it a typo?

File: .github/workflows/style.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      - name: Cache QA Test dependencies
+        uses: ./.github/actions/cache-gradle-dependencies
+
+      - name: PR labels
+        uses: joerick/pr-labels-action@v1.0.6
+
+      - name: Setup Go build environment
+        run: |
+          if [[ -n "$GITHUB_PR_LABEL_CI_RELEASE_BUILD" ]]; then
+            cci-export GOTAGS release
```
Comment (msugakov) [inline .github/workflows/style.yaml:45]: Does `cci-export` work in GHA?

**Comment by msugakov** [comment]
@gavin-stackrox 

> Any performance improvement?

:trollface: GHA 39 min, Prow 20 min.

File: .github/workflows/style.yaml (hunk 2 above)
Comment (janisz) [reply .github/workflows/style.yaml:20]: How about a script that will replace `image: quay.io/stackrox-io/apollo-ci:stackrox-test-*` with latest tag when rox images repo is tagged?

File: .github/workflows/style.yaml (hunk 6 above)
Comment (janisz) [reply .github/workflows/style.yaml:53]: I have a mixed feeling about this. In GHA we do not this wrapper and `make style` is enough.

File: .github/workflows/style.yaml (hunk 8 above)
Comment (janisz) [reply .github/workflows/style.yaml:45]: It's in our test container
```
docker run -it quay.io/stackrox-io/apollo-ci:stackrox-test-0.3.54 /bin/bash -c cci-export
Usage: /bin/real-bash KEY VALUE
```

Comment (janisz) [reply .github/workflows/style.yaml:45]: Ah, I got it. It's there but it's not working correctly :) Good catch

File: .github/workflows/style.yaml (hunk 5 above)
Comment (janisz) [reply .github/workflows/style.yaml:40]: I think this step is not required anyway but I'll fix build.yaml

File: .github/workflows/style.yaml (hunk 4 above)
Comment (janisz) [reply .github/workflows/style.yaml:17]: Let's do it

File: .github/workflows/style.yaml (hunk 5 above)
Comment (janisz) [reply .github/workflows/style.yaml]: https://github.com/stackrox/stackrox/pull/4049

File: .github/actions/cache-ui-dependencies/action.yaml [hunk 9]
```diff
@@ -13,12 +13,15 @@ runs:
       with:
         path: |
           ui/deps
+          /github/home/.cache/yarn/v6
```
Comment (porridge) [inline .github/actions/cache-ui-dependencies/action.yaml:16]: How about adding a note to wherever v6 is picked to update this line when changing?
Or strip `/v6` here?

**Review by porridge — APPROVED** [review]
LGTM but obviously we should not make this the default or required until we can reduce the execution time.

File: .github/actions/cache-ui-dependencies/action.yaml (hunk 9 above)
Comment (janisz) [reply .github/actions/cache-ui-dependencies/action.yaml:16]: I think we can drop version here so no need to update it. In theory we should obtain this path with `yarn cache dir` but then action need to have yarn installed

File: .github/workflows/style.yaml (hunk 6 above)
Comment (msugakov) [reply .github/workflows/style.yaml:41]: I'm ok if @gavin-stackrox is ok too

Comment (msugakov) [reply .github/workflows/style.yaml:41]: Well, let me correct myself. The advantage of `scripts/ci/jobs/style-checks.sh` is that it saves junit reports in a standard location which _theoretically_ makes it easier to consume and/or mass-change logic. Maybe there's a value in that?

File: .github/workflows/style.yaml (hunk 8 above)
Comment (msugakov) [reply .github/workflows/style.yaml]: Well, GOTAGS thing is gone now, but it was done before. See e.g. this revision:
https://github.com/stackrox/stackrox/blob/ba9956b42f9b90169c2ee047a7c1723ea3eb5377/.circleci/config.yml#L2249-L2255

I wonder if had any effect or was just there for consistency?

Comment (msugakov) [reply .github/workflows/style.yaml]: Reply from @janisz in DM:

> https://github.com/stackrox/stackrox/blob/master/Makefile#L137-L142 golangci runs for two tags It think other checks are not tags sensitive

**Pull request merged**